### PR TITLE
Editable ChoiceType in list doesn't support discriminators

### PR DESCRIPTION
**Pull request by willemverspyck**
When using a field "category" that consist of values of both OneCategory and TwoCategory class (both discriminator fields) and edit this field in the "list" view, select another category and click on "Ok". You will get:

Expected argument of type "?App\Entity\CategoryInterface", "string" given at property path "category".

Because the DataTransformerResolver return NULL, because the checks fails that the className and targetModel are equal: https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/Form/DataTransformerResolver.php#L80

Solution will be to check if targetModel is instance of className:

SonataAdmin:

```php
/**
 * @inheritDoc
 */
protected function configureListFields(ListMapper $listMapper): void
{
    $listMapper
        ->add('category', FieldDescriptionInterface::TYPE_CHOICE, [
            'choices' => [
                '1' => 'Name 1',
                '2' => 'Name 2',
            ],
            'class' => AbstractCategory::class,
            'editable' => true,
        ]);
}
```

Entity:

```php
#[ORM\Entity]
class OneCategory extends AbstractCategory {
}

#[ORM\Entity]
class TwoCategory extends AbstractCategory {
} 

#[ORM\DiscriminatorColumn(name: 'discriminator', type: Types::STRING, length: 256)]
#[ORM\DiscriminatorMap(value: [
    'OneCategory' => 'OneCategory',
    'TwoCategory' => 'TwoCategory',
])]
#[ORM\Entity(repositoryClass: CategoryRepository::class)]
#[ORM\InheritanceType(value: 'SINGLE_TABLE')]
#[ORM\Table(name: 'category')]
abstract class AbstractCategory {
}
```

I am targeting this branch, because BC.

## Changelog

```markdown
### Changed
Change if targetClass is instance of other className
```

## To do

I only don't know how to resolve this Psalm error:

```php
src/Form/DataTransformerResolver.php:81:64: ArgumentTypeCoercion: Argument 2 of Sonata\AdminBundle\Form\DataTransformer\ModelToIdTransformer::__construct expects class-string, parent type string provided (see https://psalm.dev/193)
```

Someone can help me with that?

### CODE REVIEW

File: src/Form/DataTransformerResolver.php [hunk 1]
```diff
@@ -77,7 +77,7 @@ public function resolve(
         if (FieldDescriptionInterface::TYPE_CHOICE === $fieldType) {
             $className = $fieldDescription->getOption('class');
```
Comment (VincentLanglet) [inline src/Form/DataTransformerResolver.php:79]:
```suggestion
            /** @var class-string|null $className */
            $className = $fieldDescription->getOption('class');
```
Could be a way to fix this.

Comment (VincentLanglet) [reply src/Form/DataTransformerResolver.php:79]: But maybe it's better to check something like `class_exists($className)`

**Review by VincentLanglet — COMMENTED** [review]
Can you add an unit test ?

File: src/Form/DataTransformerResolver.php (hunk 1 above)
Comment (willemverspyck) [reply src/Form/DataTransformerResolver.php:79]: Added the class_exists, check if method getTargetModel is not equal to NULL and the Psalm definition. Still get errors related to PHPStan and Psalm, but I don't think these are related to my changes?

**Comment by willemverspyck** [comment]
> Can you add an unit test ?

Done! Replaced stdClass with existing Entity class and added abstract class to the same Entity class

File: src/Form/DataTransformerResolver.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Form/DataTransformerResolver.php:79]: I will look at it after the merge of https://github.com/sonata-project/SonataAdminBundle/pull/7716 which fix already multiple psalm/phpstan issue

Comment (willemverspyck) [reply src/Form/DataTransformerResolver.php:79]: Thanks!

**Comment by VincentLanglet** [comment]
My PR is merged, can you rebase ? I'll take a look at your static analysis issue

**Comment by willemverspyck** [comment]
Perfect @VincentLanglet. All checks have passed now 👍

**Comment by VincentLanglet** [comment]
> Perfect @VincentLanglet. All checks have passed now 👍

Can you try without the `/** @var class-string|null $className */` annotation ?
I was expecting all the class_exists, is_a, checks made for this.

I have no error https://psalm.dev/r/1b90b5cd98

**Comment by willemverspyck** [comment]
Annotation has been removed 👍

**Review by VincentLanglet — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks

**Pull request merged**